### ISSUE TITLE
Database locked error with SqliteHuey

### ISSUE THREAD

**Issue by rajaravivarma-r** [coleifer/huey#445]
Hi nice library, I like the Sqlite backend so much since it comes pre-installed almost everywhere. My problem is I get a `database is locked` error randomly.
```
[2019-08-22 00:08:50,358] ERROR:huey.consumer.Worker:Worker-2:Error reading from queue
Traceback (most recent call last):
  File "/Users/rajaravivarma/.local/share/virtualenvs/youtube-dl-parallel-o-ngZiPY/lib/python3.7/site-packages/huey/consumer.py", line 94, in loop
    task = self.huey.dequeue()
  File "/Users/rajaravivarma/.local/share/virtualenvs/youtube-dl-parallel-o-ngZiPY/lib/python3.7/site-packages/huey/api.py", line 282, in dequeue
    data = self.storage.dequeue()
  File "/Users/rajaravivarma/.local/share/virtualenvs/youtube-dl-parallel-o-ngZiPY/lib/python3.7/site-packages/huey/storage.py", line 712, in dequeue
    with self.db(commit=True) as curs:
  File "/Users/rajaravivarma/.pyenv/versions/3.7.3/lib/python3.7/contextlib.py", line 112, in __enter__
    return next(self.gen)
  File "/Users/rajaravivarma/.local/share/virtualenvs/youtube-dl-parallel-o-ngZiPY/lib/python3.7/site-packages/huey/storage.py", line 647, in db
    if commit: cursor.execute(self.begin_sql)
sqlite3.OperationalError: database is locked
```
This is the code I am using.
```python
import re

from huey import SqliteHuey
import youtube_dl

huey = SqliteHuey(filename="downloads.db")


@huey.task(retries=2, retry_delay=60)
def download(url):
    ignorable_errors = ["HTTP Error 404", "no video in this tweet"]
    status = re.match(r"https.*/status/(\d+)", url).groups(0)[0]
    ydl_opts = {"quiet": True, "outtmpl": f"{status}.%(ext)s"}
    with youtube_dl.YoutubeDL(ydl_opts) as ydl:
        try:
            ydl.download([url])
        except Exception as e:
            error_message = str(e)
            if any(msg in error_message for msg in ignorable_errors):
                pass
            else:
                raise e
```

**Comment by coleifer** [comment]
Unfortunately that's just the nature of Sqlite. It only allows one writer at a time. Huey's sqlite storage implementation makes every effort to keep the write transactions as short as possible, so that the lock is not held any longer than necessary. So two possibilities occur to me:

* despite the short lock duration, you've still got multiple workers trying to write to the db at the same time.
* you have another process or application that's accessing downloads.db and is misbehaving with regards to how it uses transactions

You might check the journal mode of your downloads.db and make sure it was created using journal mode (`pragma journal_mode;`). Alternatively you can try deleting the db and setting the journal mode to `pragma journal_mode=delete`, but I don't think that will necessarily help.

**Comment by rajaravivarma-r** [comment]
Thanks for the quick reply. I checked the source and I didn't see a mutex (Lock) being used in the `BaseSqlStorage` class. But I will check thoroughly the whole thing when I get the time.

**Comment by coleifer** [comment]
There's no mutex -- the lock is internal to sqlite.

**Comment by rajaravivarma-r** [comment]
I read about it and it seems that one way of avoiding this error is by increasing the `timeout` a thread is waiting for the database to be unlocked. And the kwargs can be used to set a default timeout if one does not exist or use the user specified one. Would you like a PR for that change. Something like this?
```python
    def _create_connection(self):
        self._conn_kwargs = {"timeout": 5, **self._conn_kwargs}
        conn = sqlite3.connect(self.filename, **self._conn_kwargs)
```

**Comment by coleifer** [comment]
I have a hard time believing that increasing the timeout is the right answer. 5s is a lot of time! Possibly you have another process that is accessing the db?

**Comment by rajaravivarma-r** [comment]
It didn't solve the problem completely, but it did indeed reduce the number of times the exception was thrown.

No, I checked it using lsof command and no other processes access the db.

**Comment by coleifer** [comment]
You didn't mention how you are running the consumer... what type / how many workers? How much work are you throwing at it / how long is the backlog at peak?

**Comment by rajaravivarma-r** [comment]
`huey_consumer --workers 3 --worker-type thread --no-periodic huey_youtube_dl.huey` This is how I am running the worker. I have queued almost 20 thousand tasks. I am not sure about the backlog. I will check it and update the thread.

**Comment by coleifer** [comment]
Oh shit bro. You can't punish sqlite like that without getting into trouble. You know sqlite only allows one writer but when the stars align you might hit 3 writers... Use a different storage, file is a simple one if you don't want to use Redis. Or since you're just shelling out to YouTube dl, just use gevent+minihuey.

**Comment by rajaravivarma-r** [comment]
Oh. Ok. I just queued tasks once before starting the `huey_consumer`. I wanted to use Sqlite because I may use it to review the list of errors with available Sqlite Browsers, but I don't know if there are any browsers available for Redis. 

However, almost all of those tasks finished successfully after I increased the timeout.

Thanks for the information you have provided though. I will try to use the right tools in the future. :)
